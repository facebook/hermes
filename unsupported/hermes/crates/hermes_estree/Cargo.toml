[package]
name = "hermes_estree"
version = "0.1.0"
publish               = false
authors.workspace     = true
description.workspace = true
edition.workspace     = true
homepage.workspace    = true
keywords.workspace    = true
license.workspace     = true
repository.workspace  = true

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
insta = { workspace = true }
serde = { workspace = true }
serde_json = { workspace = true }
static_assertions = { workspace = true }

[build-dependencies]
hermes_estree_codegen = { workspace = true }