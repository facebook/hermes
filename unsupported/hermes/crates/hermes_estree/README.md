# hermes_estree

This crate is a Rust representation of the [ESTree format](https://github.com/estree/estree/tree/master) and
popular extenions including JSX and (eventually) Flow and TypeScript.

Because this crate is intended to support usage in both Rust toolchains and for interop in JavaSCript-based toolchains, the format is designed to support accurate serialization to/from estree-compatible JSON.
