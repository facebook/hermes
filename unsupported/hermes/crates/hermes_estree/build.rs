use hermes_estree_codegen::estree;

// Example custom build script.
fn main() {
    // Re-run if the codegen files change
    println!("cargo:rerun-if-changed=../hermes_estree_codegen/src/codegen.rs");
    println!("cargo:rerun-if-changed=../hermes_estree_codegen/src/lib.rs");
    println!("cargo:rerun-if-changed=../hermes_estree_codegen/src/ecmascript.json");
    println!("cargo:rerun-if-changed=../hermes_estree_codegen");

    let src = estree();
    std::fs::write("src/generated.rs", src).unwrap();
}
