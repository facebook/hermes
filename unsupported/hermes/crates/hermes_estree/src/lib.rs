mod binding;
mod generated;
mod generated_extensions;
mod js_value;
mod range;
mod visit;

pub use binding::{Binding, BindingId};
pub use generated::*;
pub use generated_extensions::*;
pub use js_value::{JsValue, Number};
pub use range::SourceRange;
pub use visit::*;

#[cfg(test)]
mod tests {
    use insta::{assert_snapshot, glob};

    use super::*;

    #[test]
    fn fixtures() {
        glob!("fixtures/**.json", |path| {
            println!("{:?}", path);
            let input = std::fs::read_to_string(path).unwrap();
            let ast: Program = serde_json::from_str(&input).unwrap();
            let serialized = serde_json::to_string_pretty(&ast).unwrap();
            assert_snapshot!(format!("Input:\n{input}\n\nOutput:\n{serialized}"));
        });
    }
}
