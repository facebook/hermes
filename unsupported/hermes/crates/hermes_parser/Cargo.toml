[package]
name = "hermes_parser"
version = "0.1.0"
authors.workspace = true
description.workspace = true
edition.workspace = true
homepage.workspace = true
keywords.workspace = true
license.workspace = true
repository.workspace = true

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
insta = { workspace = true }
hermes_diagnostics = { workspace = true }
hermes_estree = { workspace = true }
hermes = { workspace = true } 
juno_support = { workspace = true }
serde_json = { workspace = true }

[build-dependencies]
hermes_estree_codegen = { workspace = true }