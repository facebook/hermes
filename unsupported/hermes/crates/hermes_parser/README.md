# hermes_parser

Wrapper around the Hermes parser that exposes parse results as a `hermes_estree` AST.
