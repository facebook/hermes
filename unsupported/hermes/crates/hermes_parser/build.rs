use hermes_estree_codegen::estree_hermes;

// Example custom build script.
fn main() {
    // Re-run if the codegen files change
    println!("cargo:rerun-if-changed=../hermes_estree_codegen/src/codegen.rs");
    println!("cargo:rerun-if-changed=../hermes_estree_codegen/src/lib.rs");
    println!("cargo:rerun-if-changed=../hermes_estree_codegen/src/ecmascript.json");
    println!("cargo:rerun-if-changed=../hermes_estree_codegen");

    let src = estree_hermes();
    std::fs::write("src/generated.rs", src).unwrap();
}
