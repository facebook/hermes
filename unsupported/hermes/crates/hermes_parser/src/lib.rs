/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

include!(concat!(env!("OUT_DIR"), "/generated.rs"));
mod generated_extension;

use generated_extension::convert_comment;
use generated_extension::convert_smloc;
pub use generated_extension::Comment;
use generated_extension::Context;
use generated_extension::FromHermes;
use hermes::parser::HermesParser;
pub use hermes::parser::ParserDialect;
pub use hermes::parser::ParserFlags;
use hermes::utf::utf8_with_surrogates_to_string;
use hermes_diagnostics::Diagnostic;
use hermes_estree::Program;
use hermes_estree::SourceRange;
use juno_support::NullTerminatedBuf;

pub struct ParseResult {
    pub ast: Program,
    pub comments: Vec<Comment>,
}

pub fn parse(
    source: &str,
    _file: &str,
    flags: ParserFlags,
) -> Result<ParseResult, Vec<Diagnostic>> {
    let buf = NullTerminatedBuf::from_str_check(source);
    let result = HermesParser::parse(flags, &buf);
    let mut cx = Context::new(&buf);
    if result.has_errors() {
        let error_messages = result.messages();
        return Err(error_messages
            .iter()
            .map(|diag| {
                let message = utf8_with_surrogates_to_string(diag.message.as_slice()).unwrap();
                let start = convert_smloc(&cx, diag.loc) as u32;
                Diagnostic::invalid_syntax(
                    message,
                    SourceRange {
                        start,
                        end: start + 1,
                    },
                )
            })
            .collect());
    }

    let ast = FromHermes::convert(&mut cx, result.root().unwrap())?;
    let comments = result
        .comments()
        .iter()
        .map(|comment| convert_comment(&mut cx, comment))
        .collect();

    Ok(ParseResult { ast, comments })
}
