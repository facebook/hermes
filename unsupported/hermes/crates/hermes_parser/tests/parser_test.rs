/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use std::env;

use hermes_estree::SourceType;
use hermes_parser::ParserDialect;
use hermes_parser::ParserFlags;
use hermes_parser::parse;
use insta::assert_snapshot;
use insta::glob;

#[test]
fn fixtures() {
    let mut settings = insta::Settings::clone_current();
    if let Ok(path) = env::var("SNAPSHOT_PATH") {
        settings.set_snapshot_path(path);
    }
    let _guard = settings.bind_to_scope();
    glob!("fixtures/**.js", |path| {
        println!("fixture {}", path.to_str().unwrap());
        let input = std::fs::read_to_string(path).unwrap();
        let result = parse(
            &input,
            path.to_str().unwrap(),
            ParserFlags {
                strict_mode: true,
                enable_jsx: false,
                dialect: ParserDialect::Flow,
                store_doc_block: false,
                store_comments: true,
            },
        )
        .unwrap();
        let mut ast = result.ast;
        let comments = result.comments;
        // TODO: hack to prevent changing lots of fixtures all at once
        ast.source_type = SourceType::Script;
        let output = serde_json::to_string_pretty(&ast).unwrap();
        let output = output.trim();
        let comments_json = serde_json::to_string_pretty(&comments).unwrap();
        assert_snapshot!(format!(
            "Input:\n{input}\n\nOutput:\n{output}\n\nComments:{comments_json}"
        ));
    });
}
