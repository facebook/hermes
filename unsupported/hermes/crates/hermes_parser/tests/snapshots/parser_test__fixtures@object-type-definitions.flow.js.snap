---
source: crates/hermes_parser/tests/parser_test.rs
expression: "format!(\"Input:\\n{input}\\n\\nOutput:\\n{output}\\n\\nComments:{comments_json}\")"
input_file: crates/hermes_parser/tests/fixtures/object-type-definitions.flow.js
---
Input:
/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

// User
type User = {
    // name
    name: string,
};

type Map = {
    [string]: Value,
};

type Actor = {
    lastName: string,
    ...User,
}


type Data = {
    "hello": string,
    123: ?string,
}



Output:
{
  "type": "Program",
  "body": [
    {
      "type": "TypeAlias",
      "id": {
        "type": "Identifier",
        "name": "User",
        "typeAnnotation": null,
        "loc": null,
        "range": [
          205,
          209
        ]
      },
      "typeParameters": null,
      "right": {
        "type": "ObjectTypeAnnotation",
        "properties": [
          {
            "type": "ObjectTypeProperty",
            "key": {
              "type": "Identifier",
              "name": "name",
              "typeAnnotation": null,
              "loc": null,
              "range": [
                230,
                234
              ]
            },
            "value": {
              "type": "StringTypeAnnotation",
              "loc": null,
              "range": [
                236,
                242
              ]
            },
            "method": false,
            "optional": false,
            "static": false,
            "proto": false,
            "variance": null,
            "kind": "init",
            "loc": null,
            "range": [
              230,
              242
            ]
          }
        ],
        "indexers": [],
        "callProperties": [],
        "internalSlots": [],
        "inexact": false,
        "exact": false,
        "loc": null,
        "range": [
          212,
          245
        ]
      },
      "loc": null,
      "range": [
        200,
        246
      ]
    },
    {
      "type": "TypeAlias",
      "id": {
        "type": "Identifier",
        "name": "Map",
        "typeAnnotation": null,
        "loc": null,
        "range": [
          253,
          256
        ]
      },
      "typeParameters": null,
      "right": {
        "type": "ObjectTypeAnnotation",
        "properties": [],
        "indexers": [
          {
            "type": "ObjectTypeIndexer",
            "id": null,
            "key": {
              "type": "StringTypeAnnotation",
              "loc": null,
              "range": [
                266,
                272
              ]
            },
            "value": {
              "type": "GenericTypeAnnotation",
              "id": {
                "type": "Identifier",
                "name": "Value",
                "typeAnnotation": null,
                "loc": null,
                "range": [
                  275,
                  280
                ]
              },
              "typeParameters": null,
              "loc": null,
              "range": [
                275,
                280
              ]
            },
            "static": false,
            "variance": null,
            "loc": null,
            "range": [
              265,
              280
            ]
          }
        ],
        "callProperties": [],
        "internalSlots": [],
        "inexact": false,
        "exact": false,
        "loc": null,
        "range": [
          259,
          283
        ]
      },
      "loc": null,
      "range": [
        248,
        284
      ]
    },
    {
      "type": "TypeAlias",
      "id": {
        "type": "Identifier",
        "name": "Actor",
        "typeAnnotation": null,
        "loc": null,
        "range": [
          291,
          296
        ]
      },
      "typeParameters": null,
      "right": {
        "type": "ObjectTypeAnnotation",
        "properties": [
          {
            "type": "ObjectTypeProperty",
            "key": {
              "type": "Identifier",
              "name": "lastName",
              "typeAnnotation": null,
              "loc": null,
              "range": [
                305,
                313
              ]
            },
            "value": {
              "type": "StringTypeAnnotation",
              "loc": null,
              "range": [
                315,
                321
              ]
            },
            "method": false,
            "optional": false,
            "static": false,
            "proto": false,
            "variance": null,
            "kind": "init",
            "loc": null,
            "range": [
              305,
              321
            ]
          },
          {
            "type": "ObjectTypeSpreadProperty",
            "argument": {
              "type": "GenericTypeAnnotation",
              "id": {
                "type": "Identifier",
                "name": "User",
                "typeAnnotation": null,
                "loc": null,
                "range": [
                  330,
                  334
                ]
              },
              "typeParameters": null,
              "loc": null,
              "range": [
                330,
                334
              ]
            },
            "loc": null,
            "range": [
              327,
              334
            ]
          }
        ],
        "indexers": [],
        "callProperties": [],
        "internalSlots": [],
        "inexact": false,
        "exact": false,
        "loc": null,
        "range": [
          299,
          337
        ]
      },
      "loc": null,
      "range": [
        286,
        337
      ]
    },
    {
      "type": "TypeAlias",
      "id": {
        "type": "Identifier",
        "name": "Data",
        "typeAnnotation": null,
        "loc": null,
        "range": [
          345,
          349
        ]
      },
      "typeParameters": null,
      "right": {
        "type": "ObjectTypeAnnotation",
        "properties": [
          {
            "type": "ObjectTypeProperty",
            "key": {
              "type": "StringLiteral",
              "value": "hello",
              "loc": null,
              "range": [
                358,
                365
              ]
            },
            "value": {
              "type": "StringTypeAnnotation",
              "loc": null,
              "range": [
                367,
                373
              ]
            },
            "method": false,
            "optional": false,
            "static": false,
            "proto": false,
            "variance": null,
            "kind": "init",
            "loc": null,
            "range": [
              358,
              373
            ]
          },
          {
            "type": "ObjectTypeProperty",
            "key": {
              "type": "NumericLiteral",
              "value": 123.0,
              "loc": null,
              "range": [
                379,
                382
              ]
            },
            "value": {
              "type": "NullableTypeAnnotation",
              "typeAnnotation": {
                "type": "StringTypeAnnotation",
                "loc": null,
                "range": [
                  385,
                  391
                ]
              },
              "loc": null,
              "range": [
                384,
                391
              ]
            },
            "method": false,
            "optional": false,
            "static": false,
            "proto": false,
            "variance": null,
            "kind": "init",
            "loc": null,
            "range": [
              379,
              391
            ]
          }
        ],
        "indexers": [],
        "callProperties": [],
        "internalSlots": [],
        "inexact": false,
        "exact": false,
        "loc": null,
        "range": [
          352,
          394
        ]
      },
      "loc": null,
      "range": [
        340,
        394
      ]
    }
  ],
  "sourceType": "script",
  "loc": null,
  "range": [
    200,
    394
  ]
}

Comments:[
  {
    "value": "*\n * Copyright (c) Meta Platforms, Inc. and affiliates.\n *\n * This source code is licensed under the MIT license found in the\n * LICENSE file in the root directory of this source tree.\n ",
    "range": [
      0,
      190
    ]
  },
  {
    "value": " User",
    "range": [
      192,
      199
    ]
  },
  {
    "value": " name",
    "range": [
      218,
      225
    ]
  }
]
